### PR TITLE
Add a way to use custom coap endpoint for LeshanServer

### PR DESCRIPTION
**Pull request by sbernard31**
It should solve #372.

### CODE REVIEW

**Comment by davidgraeff** [comment]
Lgtm. Just some question: Is it actually necessary to provide all SSL configuration parameters to the builder, if the endpoint factory will return an own dtlsConnector? Is there a specific order for the builder methods now?

**Comment by sbernard31** [comment]
There is no order. All is assembled in right order on `build()` method.
All the coap/dtls configuration done by the builder should be used when you create your endpoint in the factory. (I mean all parameters of this 2 methods)
```java
// For example 
CoapEndpoint createUnsecuredEndpoint(InetSocketAddress address, NetworkConfig coapConfig, ObservationStore store){
       return new CoapEndpoint(localAddress, coapConfig, store); 
}
CoapEndpoint createSecuredEndpoint(DtlsConnectorConfig dtlsConfig, NetworkConfig coapConfig,
            ObservationStore store) {
       return new CoapEndpoint(new DTLSConnector(dtlsConfig), coapConfig, store, null);
}
```
You're free to not using it but you must know what you are doing. E.g. create a "DTLS endpoint" without using our PskStore may break PSK authentication.

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java [hunk 1]
```diff
@@ -157,7 +156,7 @@ protected Resource createRoot() {
                 return new RootResource();
             }
         };
-        nonSecureEndpoint = new CoapEndpoint(localAddress, coapConfig, this.observationService.getObservationStore());
+        nonSecureEndpoint = unsecuredEndpoint;
```
Comment (msangoi) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java]: you can rename it to unsecureEndpoint.

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanBootstrapServer.java [hunk 2]
```diff
@@ -47,28 +45,28 @@
...
         Validate.notNull(coapConfig, "coapConfig must not be null");
-        Validate.notNull(dtlsConfig, "dtlsConfig must not be null");
 
         this.bsStore = bsStore;
         this.bsSecurityStore = bsSecurityStore;
 
         // init CoAP server
         coapServer = new CoapServer(coapConfig);
-        nonSecureEndpoint = new CoapEndpoint(localAddress, coapConfig);
+        nonSecureEndpoint = unsecuredEndpoint;
```
Comment (msangoi) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanBootstrapServer.java]: unsecuredEndpoint

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java [hunk 3]
```diff
@@ -269,6 +274,15 @@ public LeshanServerBuilder setDtlsConfig(DtlsConnectorConfig.Builder config) {
     }
 
     /**
+     * Used to create custom CoAP endpoint, this is only for advanced users. <br>
+     * DTLSConnector is expected for secured endpoint.
```
Comment (msangoi) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java]: We should explain here that localAddress/dtlsConfig are ignored if an endpoint factory is provided.

**Comment by msangoi** [comment]
👍 LGTM

**Comment by sbernard31** [comment]
Integrated in master (95566f295b516780ad8e71cecaa69e55db7a1909, 1f630498acdec9655fb2724436198103fe666dc0)

Same thing should be done for client but I need access to DTLSConnector and so Californium 2.0.0M5 which should be available soon.

**Comment by davidgraeff** [comment]
Do you have an ETA for a (pre)release? We would need a version at some point for Eclipse Smarthome.

**Comment by sbernard31** [comment]
We are waiting for Californium 2.0.0M5, once it will be available. We need to integrate some works based which depends of this 2.0.0M5 version in Leshan, then we will release a 1.0.0M4. I hope this will happen soon. (something like 1 week)

**Comment by sbernard31** [comment]
@davidgraeff I totally forgot to tell you that the [1.0.0-M4](https://github.com/eclipse/leshan/releases/tag/leshan-1.0.0-M4) is out. It integrates this feature and some other new cool bugs ;)

**Comment by davidgraeff** [comment]
Thank you :)

**Pull request closed without merging**